### ISSUE TITLE
Allow math

### ISSUE THREAD

**Issue by lindeloev** [AvatarHurden/keypirinha-currency#5]
It would be great to allow for search terms like `2155*12 eur in usd`, say if you know your monthly salary in euros and want to know the yearly salary in dollars. The same goes for `+`, `-` and `/`. The next level would be support for `^` and parentheses, which I also sometimes find myself using. So the ultimate test would be if the following worked:

`2*(3+4) - 8/2^2 usd in eur`

Thanks for this great package! Up until now I've used Google Search for these conversions, so this is the feeling I'd like to emulate in Keypirinha.

A smaller wish for first-time users is the ability to type in the currency names (dollars, euros, kroner), though that can be ambivalent. You quickly become accustomed to the abbreviations, so it's not that big of a deal, but it would make the learning curve even flatter!

**Comment by AvatarHurden** [comment]
Hi @lindeloev!

To support mathematical expressions, I think the best approach would be to use an existing parsing/eval library, as building one from scratch seems unnecessary. This would definitely be useful, and I was thinking the best approach would be something along the lines of:

`(expr curr) +/- (expr curr) +/- ... in curr`

This would allow adding or subtracting in multiple currencies (something that I actually do quite a lot). I don't think it makes sense to allow any operation besides +/- between currencies (after all, what does it mean to multiply euros by dollars?), but the `expr` for each currency could be anything (exponentiation or even `sin` and `cos`). I might tackle this when I have more time at the end of the year, but you're also free to give it a try and make a pull request.

As for adding currency names, that is something that I tried to implement before, but couldn't find a good way. If the user has to match the currency name exactly, typing `dollar` wouldn't actually give you any results (since the currency is called 'United States Dollar'). If I try to match it to the closest match using a simple algorithm, some results would be strange (again, `dollar` would match to one of the many other currencies called dollar in the world).

The best approach I could think of was to construct the list of currencies manually, but again this would bring some issues. I think almost everyone can agree that `dollar` should refer to the US dollar, but what about `peso` (mexican, uruguayan, argentine)? In the end, I decided that the added complexity of adding the currency names was a bit too much (but, again, feel free to implement something and create a pull request!)

**Comment by AvatarHurden** [comment]
I really like this approach. I think I would prefer to add some default aliases for `dollar`, `euro` and `British pound` (the default currencies in the plugin), just to make it easier for novice users to understand.

I'll try to implement this for the next version!

**Comment by polyvertex** [comment]
> I think I would prefer to add some default aliases [...]

Just to elaborate (i.e. not trying to convince you!), the point behind not pre-defining aliases is to avoid any potential confusion. For that reason I would probably expansively emphasize both in the README and the inline documentation of the config file, that it is possible to define aliases if desired. I would also put a detailed example in comment.

Anyways, have fun! :)

**Comment by DrorHarari** [comment]
I agree with @polyvertex about having no pre defined aliases - unless you confer with the Windows regional settings and the $ (or dollar) as the local one. This will be much more work though...


>> ...I would not be happy as an Australian citizen...
Didn't know you are but now it makes sense. You Australians write great programs (KP, Everything, The V File Viewer, FastStone Capture, Atlassian, among others)

**Comment by polyvertex** [comment]
> [...] unless you confer with the Windows regional settings [...] This will be much more work though...

Not necessarily I'd say! The official Calc package reads system's regional settings and it's not that painful. Almost straightforward :)

**Comment by DrorHarari** [comment]
Not a disappointment. I'd always take a great software.

**Comment by AvatarHurden** [comment]
Hey @polyvertex @lindeloev!
I've released a new version of the plugin, now with support for math and aliases (and a much more flexible syntax). I didn't manage to include regional settings for default currencies, but I looked a bit into it and there is some nice information that I could use in the future (default currency symbol, number of decimal places, etc).

Let me know what you think!

**Comment by DrorHarari** [comment]
@AvatarHurden - great addition to a great plugin. 

A useful future addition would be 
`{expr} {'+'|'-'} {number}'%'`
To allow adding tips, calculating discounts etc.
So `12+2 +17% usd` may give the equivalent of (12+2)*1.17

**Comment by DrorHarari** [comment]
@lindeloev I agree that this seem in contrast to the order of operations but on the other hand it is very intuitive and the syntax of {expr} {'+'|'-'} {number}'%' is consistent (note that this can only appear at the end). Forcing a tab with special option is, as you mentioned, an overkill.

BTW, the {number}'%' could have an alias defined in the config so that if you write `12+2 +VAT usd` it will be the equivalent of `(12+2)*1.135 usd` in some state where the VAT is 13.5%

**Comment by DrorHarari** [comment]
Having looked at the code, I can see that Currency's syntax is already not your pure-school arithmetic-order-of-operations syntax but rather a more useful one. This allows for expressions like: `10 usd + 100 jpy * 2` which is evaluated as `(10 usd + 100 jpy) * 2`

To support the above % idea, the syntax of:
```
prog := sources (to_key destinations)? extra?
...
extra := ('+' | '-' | '*' | '/' | '**' | '^' ) expr
...
number := (0|[1-9][0-9]*)([.,][0-9]+)?([eE][+-]?[0-9]+)?
```
should change to:
```
prog := sources (to_key destinations)? extra?
...
extra := ('+' | '-' | '*' | '/' | '**' | '^' ) ( expr | ('+' | '-' | '*' | '/')  number '%')
...
number := (0|[1-9][0-9]*)([.,][0-9]+)?([eE][+-]?[0-9]+)?
```

**Comment by AvatarHurden** [comment]
@lindeloev I don't really know how `PackageControl` works, but I think one way to verify what version you have is to open the package configuration file. If it has the `aliases` section, then it should be the newest version.

@DrorHarari This looks like a good idea. The only issue is that your suggestion accepts something like `10 usd + 10%`, but not `10 + 10% usd`. I think the best scenario is one in which you can specify a percentage for each individual expression. My suggestion is to change `expr` to 
``` 
expr := add_expr (('+' | '-') number '%')?
```
This would allow more complex operations, such as `(10 + 10%) usd + (100 + 9%) jpy`, which could be useful for different taxes (or something along those lines). What do you think?

BTW, I don't think adding multiplication and division to percentages is a good idea, as it would be confusing (are you multiplying by 10% or by 10% of the value?)

**Comment by DrorHarari** [comment]
@AvatarHurden as I said, I was not fully aware of all the flexible options you offer in Currency so I wanted to be conservative on that to make sure it is clear to what the % is added.  I think your option is more useful.

As for the multiplication, I thought of it as the common sense of meaning `10% = 0.1` or in general `x%` meaning `x/100` (treated as a pure number) so it's a useful and clear shortcut. As for division, it got in there from symmetry considerations😎  -  clearly one can't multiply by `10%` of the value as then you'll get squared coins which are not practical (which is why I am a bit puzzled by the `**` in the context of currencies).

**Comment by DrorHarari** [comment]
> It will be nice in the context of % support to also have the ability to have 'named percentages'.

To clarify: the idea was to be able to configure the name VAT (for example) to be 1.17 such that any time I type "12 USD VAT" it will multiply the result by 1.17. The VAT (or other 'tax codes' or 'deal codes') would be treated as 'currency modifiers'. Thus, they would not affect the arithmetic part of the syntax...
